Bump pyo3 version

surfingreg opened the pull request:
## Status

Ready for review

## Description of Changes

- Fixes #7277
- Upgrades pyo3 crate in redwood/Cargo.toml from 0.18.0 to 0.23.4
- Transitions to pyo3's bound API style 
- Relevant references:  
-  [PyO3 Migration Guide](https://pyo3.rs/v0.23.4/migration.html?highlight=pybytes#to-python-conversions-changed-for-byte-collections-vec-u8-n-and-smallvec)
- https://docs.rs/pyo3/latest/pyo3/struct.Bound.html

## Testing

How should the reviewer test this PR?
_- Standard:_
```
Fastest/narrowest test of changed Rust code:
cd securedrop
cargo test --package redwood

Standard dev/docker environment:
cd securedrop
make dev                                               # run development servers
make test
```

Write out any special testing steps here.
_None required_

## Deployment

Any special considerations for deployment? Consider both:

1. Upgrading existing production instances.
_I don't think so._

2. New installs.
_No._

## Checklist

### If you made changes to the server application code:

- [ ] Linting (`make lint`) and tests (`make test`) pass in the development container

_"make test" fails on some [ar] locales. Is this to be expected? Is there anything I can do to correct the test locally? It passes all others._

### If you made changes to `securedrop-admin`:

- [ ] Linting and tests (`make -C admin test`) pass in the admin development container

### If you made changes to the system configuration:

- [ ] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass

### If you added or removed a file deployed with the application:

- [ ] I have updated AppArmor rules to include the change

### If you made non-trivial code changes:

- [ ] I have written a test plan and validated it for this PR

_Not applicable. Existing tests suffice._

Choose one of the following:

- [ ] I have opened a PR in the [docs repo](https://github.com/freedomofpress/securedrop-docs) for these changes, or will do so later
- [X] I would appreciate help with the documentation
- [ ] These changes do not require documentation

### If you added or updated a reference to a production code dependency:

Production code dependencies are defined in:
- `admin/requirements.in`
- `admin/requirements-ansible.in`
- `securedrop/requirements/python3/requirements.in`
- `securedrop/requirements/python3/translation.in` (used in the build
  container)

_I'm not sure. I changed one Rust dependency in redwood/Cargo.toml. It passes cargo vet and cargo audit._

If you changed another `requirements.in` file that applies only to development
or testing environments, then no diff review is required, and you can skip
(remove) this section.

Choose one of the following:

- [ ] I have performed a diff review and pasted the contents to [the packaging wiki](https://github.com/freedomofpress/securedrop-debian-packaging/wiki)
- [X] I would like[/recommend] someone else to do the diff review [since I'm brand new here]
- [ ] I am silencing an alert related to a production dependency, because (please explain below):

Commit:
Bump pyo3 version in redwood from 0.18.0 to 0.23.4

* Adapted to use the new `Bound` API
* Added explicit Python function signatures since pyo3 is getting
  rid of the implicit Option defaults.

legoktm commented:
Thanks for submitting @surfingreg; the Python CI all looks good \o/. The rust checks are flagging that `rustfmt` has some changes to make if you want to run `cargo fmt`.

The cargo-vet review part will be handled by us maintainers, so don't worry about that. I do just want to set expectations that our current focus is on getting [SecureDrop 2.12.0](https://github.com/freedomofpress/securedrop/milestone/89)  out the door; I/we will plan to get to this for the 2.13.0 release cycle.

surfingreg commented:
Cargo fmt done.

cfm commented:
Thanks, @surfingreg!  This is currently blocked on our side on freedomofpress/securedrop-engineering#112, but we'll get back to you soon.

legoktm committed:
Upgrade pyo3 to 0.24.1 for RUSTSEC-2025-0020

legoktm committed:
Update cargo-vet for new pyo3 version

* Trust David Hewitt per
<https://github.com/freedomofpress/securedrop-engineering/issues/112>.
* Trust portable-atomic and target-lexicon crates, released by Rust
Project members.
* Audit memoffset 0.9.0 -> 0.9.1

legoktm committed:
Upgrade openssl for RUSTSEC-2025-0022

legoktm commented on redwood/src/lib.rs:
```diff
@@ -124,7 +124,10 @@ pub fn is_valid_secret_key(input: &str, passphrase: String) -> Result<String> {
 /// be written to `destination`.
-#[pyfunction]
+///
+/// This macro signature addresses a new pyo3 warning about default values...
```
I'm going to move this comment to the commit message since it describes why the change was made, but doesn't need to be kept in the code itself.

legoktm commented:
Thanks again for this @surfingreg. I rebased your initial PR, and then tacked on a further upgrade to 0.24.1, the necessary cargo vet config + reviews, plus an openssl bump.

I've reviewed the Rust code you modified, it all LGTM to me, just waiting on CI now.

legoktm reviewed (approved):
:shipit:
